File: Assignment2/Maximum Advertising Revenue/dot_product.py
#Uses python3

import sys

def max_dot_product(a, b):

    # write your code here

    # sort the two arrays using bubble sort 
    for i in range(len(a)):

        # flag for swapping
        swapped =  False

        for j in range(0, len(a)-i-1):
            
            #------------------------ Sorting Array a -----------------------#

            # swap the element of array a if it's larger than the next element
            if a[j] > a[j+1]:

                # make the swap
                a[j], a[j+1] = a[j+1], a[j]

                # set the flag swapped to true
                swapped = True


            #------------------------ Sorting Array b -----------------------#
            
            # swap the element of array b if it's larger than the next element
            if b[j] > b[j+1]:

                # make the swap
                b[j], b[j+1] = b[j+1], b[j]

                # set the flag swapped to true
                swapped = True
        
        # break the loop if there is no swapped elements
        # this is for optimizing the bubble sort to give better results,
        # if the array is sorted so we don't do the operation in O(N^2)
        if swapped == False:
            break
    
    # result of dot product
    res = 0

    # multiply each element in a with the corresponding element in b to
    # maximize the dot product
    for i in range(n):
        res += a[i] * b[i]

    return res

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    a = data[1:(n + 1)]
    b = data[(n + 1):]
    print(max_dot_product(a, b))
    

File: Assignment 2/Maximum Advertising Revenue/dot_product.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <bits/stdc++.h>

using std::vector;

void swap(int &a, int &b){
    int temp = a;
    a = b;
    b = temp;
}


long long max_dot_product(vector<int> a, vector<int> b) {
    // write your code here

    // flag for swapping
    bool swapped;

    // sort the two arrays using bubble sort
    for (int i = 0; i < a.size() - 1; i++) {
        swapped = false;

        //------------------------ Sorting Array a -----------------------//

        // swap the element of array a if it's larger than the next element
        for (int j = 0; j < b.size() - i - 1; j++) {
            if (a[j] > a[j + 1]) {
                swap(a[j], a[j + 1]);        // make the swap
                swapped = true;                   // set the flag swapped to true
            }

        //------------------------ Sorting Array a -----------------------//

        //swap the element of array b if it's larger than the next element
            if (b[j] > b[j + 1]) {
                swap(b[j], b[j + 1]);       // make the swap
                swapped = true;                   // set the flag swapped to true
            }
        }

        /*
         * break the loop if there is no swapped elements
         * this is for optimizing the bubble sort to give better results,
         * if the array is sorted so we don't do the operation in O(N^2)
         */

        if (!swapped)
            break;
    }

    // result of dot product
    long long result = 0;
    for (size_t i = 0; i < a.size(); i++) {
        result += ((long long) a[i]) * b[i];
    }
    return result;
}

int main() {
    size_t n;
    std::cin >> n;
    vector<int> a(n), b(n);
    for (size_t i = 0; i < n; i++) {
        std::cin >> a[i];
    }
    for (size_t i = 0; i < n; i++) {
        std::cin >> b[i];
    }
    std::cout << max_dot_product(a, b) << std::endl;
}
